File: app/controllers/search_controller.rb
class SearchController < ApplicationController
  # Sets @show_sidebar = true before any action runs, same pattern as ChatsController
  before_action :enable_sidebar

  def index
    # Grab the search query from the URL params (e.g. /search?q=hello) and remove whitespace
    query = params[:q].to_s.strip

    # Only run the query if the user typed at least 2 characters
    if query.length >= 2

      # Search chats belonging to the current user where the title matches the query
      # ILIKE is case-insensitive LIKE in PostgreSQL
      # % means "anything before or after the query"
      @chats = current_user.chats
        .where("title ILIKE ?", "%#{query}%")
        .limit(5) # max 5 results per category

      # Search messages where the chat belongs to the current user
      # searching both content and name fields on the message
      # joins(chat: :user) lets us filter by the user through the chat association
      @messages = Message.joins(chat: :user)
      .where(users: { id: current_user.id })
      .where("messages.content ILIKE ? OR messages.name ILIKE ?", "%#{query}%", "%#{query}%")
      .limit(5)

      # Search projects belonging to the current user by name or description
      @projects = current_user.projects
        .where("name ILIKE ? OR description ILIKE ?", "%#{query}%", "%#{query}%")
        .limit(5)
    end

    # Only respond to JSON requests (called by our Stimulus controller via fetch)
    respond_to do |format|
      format.json {
        render json: {
          # Map each chat to a plain hash with only the fields the frontend needs
          chats: @chats&.map { |c| { id: c.id, title: c.title, type: "chat", url: chat_path(c) } },

          # truncate(80) cuts the message content to 80 characters so it fits in the dropdown
          # we show name if it exists, otherwise fall back to content
          messages: @messages&.map { |m| { id: m.id, content: m.content.truncate(80), name: m.name, type: "message", url: chat_path(m.chat) } },

          projects: @projects&.map { |p| { id: p.id, name: p.name, type: "project", url: project_path(p) } }
        }
      }
    end
  end

  private

  # Sets @show_sidebar = true so application.html.erb renders the sidebar for this page
  def enable_sidebar
    @show_sidebar = true
  end
end

File: app/javascript/controllers/search_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  // Declare the HTML elements this controller needs to reference
  // Stimulus will automatically find elements with data-search-target="input" etc.
  static targets = ["input", "results", "dropdown"]

  connect() {
    // Called when the controller is attached to the DOM
    // Initialize the debounce timer as null so we can use it later
    this.debounceTimer = null
  }

  toggle(e) {
    // Stop the click from bubbling up to the window listener (which would immediately close it)
    e.stopPropagation()

    // Add or remove the "open" class on the dropdown to show/hide it
    this.dropdownTarget.classList.toggle("open")

    // If we just opened it, move focus to the input so the user can type straight away
    if (this.dropdownTarget.classList.contains("open")) {
      this.inputTarget.focus()
    }
  }

  close() {
    // Called when the user clicks anywhere on the window outside the dropdown
    this.dropdownTarget.classList.remove("open")
  }

  stopClose(e) {
    // Stop clicks inside the dropdown from bubbling up to the window close listener
    e.stopPropagation()
  }

  search() {
    // Clear any previous debounce timer so we don't fire multiple requests
    clearTimeout(this.debounceTimer)

    // Get the current value of the search input
    const query = this.inputTarget.value.trim()

    // If the query is too short, reset the results and bail out
    if (query.length < 2) {
      this.resultsTarget.innerHTML = '<p class="search-hint">Type at least 2 characters...</p>'
      this.previewTarget.classList.remove("open")
      return
    }

    // Wait 300ms after the user stops typing before firing the request
    // This avoids sending a request on every single keystroke
    this.debounceTimer = setTimeout(() => {
      fetch(`/search?q=${encodeURIComponent(query)}`, {
        // Tell Rails we want JSON back, not HTML
        headers: { "Accept": "application/json" }
      })
      .then(res => res.json()) // Parse the response as JSON
      .then(data => this.renderResults(data)) // Pass the data to renderResults
    }, 300)
  }

  renderResults(data) {
      console.log(data) // ADD THIS — shows the full JSON response in browser console
    // Combine all result types into one flat array
    const all = [
      ...(data.chats || []),
      ...(data.messages || []),
      ...(data.projects || [])
    ]

    // If nothing came back, show a "no results" message
    if (all.length === 0) {
      this.resultsTarget.innerHTML = '<p class="search-hint">No results found.</p>'
      return
    }

    // Human-readable labels for each result type shown above the result title
    const labels = { chat: "Chat", message: "Message", project: "Project" }

    // Build the HTML for each result and inject it into the results container
    this.resultsTarget.innerHTML = all.map(item => `
      <div class="search-result-item"
           data-url="${item.url}"
           data-action="click->search#selectResult">
        <span class="search-result-type">${labels[item.type]}</span>
        <!-- show name if available, otherwise content, otherwise title -->
        <p class="search-result-title">${item.name || item.title || item.content}</p>
      </div>
    `).join("") // join("") turns the array of strings into one big HTML string
  }

  selectResult(e) {
  // Navigate directly to the result when clicked
  window.location.href = e.currentTarget.dataset.url
  }
}
